jetty-servlet/src/test/java/org/eclipse/jetty/servlet/ServletUpgradeTest.java

```diff
@@ -0,0 +1,352 @@
+                if (passed1 = compareString(expectedResponse1, sb.toString()))
+                {
+                    LOG.info("==============Received first expected response!" + CRLF);
+                    receivedFirstMessage = true;
+                }
+                if (passed2 = compareString(expectedResponse2, sb.toString()))
+                {
+                    LOG.info("==============Received second expected response!" + CRLF);
+                    receivedSecondMessage = true;
+                }
+                if (passed3 = compareString(expectedResponse3, sb.toString()))
+                {
+                    LOG.info("==============Received third expected response!" + CRLF);
+                    receivedThirdMessage = true;
+                }
+                LOG.info("receivedFirstMessage : " + receivedFirstMessage);
+                LOG.info("receivedSecondMessage : " + receivedSecondMessage);
+                LOG.info("receivedThirdMessage : " + receivedThirdMessage);
+                if (receivedFirstMessage && receivedSecondMessage && receivedThirdMessage)
+                {
+                    break;
+                }
+            }
+        }
+        finally
+        {
+            try
+            {
+                if (input != null)
+                {
+                    LOG.info("Closing input...");
+                    input.close();
+                    LOG.info("Input closed.");
+                }
+            }
+            catch (Exception ex)
+            {
+                LOG.error("Failed to close input:" + ex.getMessage(), ex);
+            }
+
+            try
+            {
+                if (output != null)
+                {
+                    LOG.info("Closing output...");
+                    output.close();
+                    LOG.info("Output closed .");
+                }
+            }
+            catch (Exception ex)
+            {
+                LOG.error("Failed to close output:" + ex.getMessage(), ex);
+            }
+
+            try
+            {
+                if (s != null)
+                {
+                    LOG.info("Closing socket..." + CRLF);
+                    s.close();
+                    LOG.info("Socked closed.");
+                }
+            }
+            catch (Exception ex)
+            {
+                LOG.error("Failed to close socket:" + ex.getMessage(), ex);
+            }
+        }
+
+        assertTrue(passed1 && passed2 && passed3);
+    }
+
+    private static class TestServlet extends HttpServlet
+    {
+        public void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException
+        {
+            if (request.getHeader("Upgrade") != null)
+            {
+                response.setStatus(101);
+                response.setHeader("Upgrade", "YES");
+                response.setHeader("Connection", "Upgrade");
+                request.upgrade(TestHttpUpgradeHandler.class);
```

Comment:
`upgrade()` returns the handler, I would test that it's not null and of the type I expect.

Reply:
Ok, let me add that check.